cluster-controller/src/main/java/io/strimzi/controller/cluster/operations/cluster/KafkaClusterOperations.java

```diff
@@ -39,681 +39,183 @@
-            StatefulSet ss = statefulSetOperations.get(namespace, ZookeeperCluster.zookeeperClusterName(name));
-            return new ClusterOperation<ZookeeperCluster>(ZookeeperCluster.fromStatefulSet(ss, namespace, name), null);
-        }
+        return CompositeFuture.join(result).map((Void) null);
     };
 
-    private final CompositeOperation<TopicController> deleteTopicController = new CompositeOperation<TopicController>() {
-
-        @Override
-        public String operationType() {
-            return OP_DELETE;
-        }
-
-        @Override
-        public String clusterType() {
-            return CLUSTER_TYPE_TOPIC_CONTROLLER;
-        }
-
-        @Override
-        public Future<?> composite(String namespace, ClusterOperation<TopicController> clusterOp) {
-            TopicController topicController = clusterOp.cluster();
-            return deploymentOperations.delete(namespace, topicController.getName());
-        }
-
-        @Override
-        public ClusterOperation<TopicController> getCluster(String namespace, String name) {
-            Deployment dep = deploymentOperations.get(namespace, TopicController.topicControllerName(name));
-            return new ClusterOperation<>(TopicController.fromDeployment(namespace, name, dep), null);
-        }
+    private final Future<Void> updateZk(String namespace, String name) {
+        ConfigMap zkConfigMap = configMapOperations.get(namespace, name);
+        ZookeeperCluster zk = ZookeeperCluster.fromConfigMap(zkConfigMap);
+        Service service = zk.generateService();
+        Service headlessService = zk.generateHeadlessService();
+        Future<Void> chainFuture = Future.future();
+        zkSetOperations.scaleDown(namespace, zk.getName(), zk.getReplicas())
+                .compose(scale -> serviceOperations.reconcile(namespace, zk.getName(), service))
+                .compose(i -> serviceOperations.reconcile(namespace, zk.getHeadlessName(), headlessService))
+                .compose(i -> configMapOperations.reconcile(namespace, zk.getMetricsConfigName(), zk.generateMetricsConfigMap()))
+                .compose(i -> zkSetOperations.reconcile(namespace, zk.getName(), zk.generateStatefulSet(isOpenShift)))
+                //.compose(i -> zkSetOperations.rollingUpdate(namespace, zk.getName()))
+                .compose(diffs -> {
+                    if (diffs instanceof ReconcileResult.Patched
+                            && ((ReconcileResult.Patched<Boolean>) diffs).differences()) {
+                        return zkSetOperations.rollingUpdate(namespace, zk.getName());
+                    } else {
+                        return Future.succeededFuture();
+                    }
+                })
+                .compose(i -> zkSetOperations.scaleUp(namespace, zk.getName(), zk.getReplicas()))
+                .compose(scale -> serviceOperations.endpointReadiness(namespace, service, 1_000, operationTimeoutMs))
```

Comment:
You check the Endpoint readiness ... but it looks like you lost the Pod readiness. I think that the Endpoint readiness will complete when one pod is ready. Not when all of them are. So I think we need to go through the pods and check their readiness as before. Or did you moved the Pod readiness into some lower class?

Reply:
Ah, yeah. Pod readiness is checked on SS creation, and as part of the SS rolling update: See `StatefulSetOperations`.

The reason it's done differently for endpoints is that we need to check those after the rolling update (or SS creation). So putting them in with the `ServiceOperations` (because Services "own" Endpoints) doesn't work.

Reply:
Ok. Makes sense. I was looking for it in `KafkaSetOperationa` and `ZookeeperSetOperations`. But didn't realized it is even below it.,